### DISCUSSION TITLE
Magic `from()` methods are called only after validation when `validateAndCreate()` is invoked. So it's always marked as invalid.

### DISCUSSION THREAD

**Discussion by raveren** [spatie/laravel-data#406] [Q&A]
```php
<?php

namespace Tests\Unit;

use Spatie\LaravelData\Data;
use Tests\TestCase;

class Example3Test extends TestCase
{
    public function test_magic_from(): void
    {
        $songName = '[[ SONG NAME ]]';
        $song     = Song::from(['title' => $songName]);

        // Test begins here
        dump('first');
        $t = HitSong::from($song); // magic method `fromS` invoked as evident in output below
        dump($t->title_of_hit_song);
        self::assertEquals($songName, $t->title_of_hit_song);

        dump('second');
        $t = HitSong::validateAndCreate($song); // magic method not invoked, the title_of_hit_song is null
        dump($t->title_of_hit_song);
        self::assertEquals($songName, $t->title_of_hit_song);
    }
}

class Song extends Data
{
    public string $title;
}

class HitSong extends Data
{
    public string $title_of_hit_song;

    public static function fromS(Song $s): self
    {
        dump('custom method called');
        $o                    = new self();
        $o->title_of_hit_song = $s->title;

        return $o;
    }
}
```

output:
```
"first" // tests/Feature/asterisk/AsteriskTest.php:74
"custom method called" // tests/Feature/asterisk/AsteriskTest.php:97
"[[ SONG NAME ]]" // tests/Feature/asterisk/AsteriskTest.php:76
"second" // tests/Feature/asterisk/AsteriskTest.php:79

Illuminate\Validation\ValidationException : title of hit song is required
 /var/www/vendor/laravel/framework/src/Illuminate/Support/helpers.php:327
 /var/www/vendor/laravel/framework/src/Illuminate/Validation/Validator.php:495
 /var/www/vendor/spatie/laravel-data/src/Concerns/ValidateableData.php:30
 /var/www/vendor/spatie/laravel-data/src/Concerns/ValidateableData.php:52
```

So the only difference from issue #394 code is: `public string $title_of_hit_song;` - it's no longer optional. 

It has to be present, and the custom `from` method sets it properly. However, because the validation is launched first, it never comes to that, effectively NEVER calling the magic method and ALWAYS throwing validation exception, making it IMPOSSIBLE for the magic `from` method to be invoked via `validateAndCreate` if any validation rules are present.

**Comment by rubenvanassche** [comment]
I don't see this as an issue, the method is called validateAndCreate so yeah first it validates and then it creates an object.

Though indeed you could first create an object with a magic method and then validate it. This might not be true for all cases, since we actually see magic methods as a last resort. If everything goes well laravel-data doesn't need them.

First creating an object and then validating it is extremely dangerous because PHP could start throwing exceptions with wrong property types.

I think you could add a createAndValidate method in this case, since you're able to get the validation rules after the object is created and you could manually validate the object. But I don't think this method needs to be in the package since validation is now already such complicated topic. Adding a whole new way to do it makes it even more complicated.

**Comment by raveren** [comment]
Hello, I hope you had a long, relaxing weekend!

I am sorry, but you are viewing this matter from a maintainers perspective, and I am creating a user issue.

**The objective fact here is the library does not work as expected**. 

It is not my opinion, not even a matter of opinion.

You might not have the motivation or resources to fix it now, it's understandable, but if you expect to run with the library in the current state, it WILL be a huge issue once more people start to use it.

Talking purely from a user-of-your-functionality point of view, It's surprising (and _undocumented_) that:

1. First of all, `from()` does not perform validation. It does all casts and name resolutions, but inexplicably not validation.
2. The **ONLY way to validate** passed data is supposed to be createAndValidate, and it is utterly **broken**! It doesn't actually validate because it does not call the appropriate `from()` method, so e.g. if you're converting an object to dto there is actually *NO way to run validation*.
3. The third surprising thing is that Spatie, an organization with such great reputation in the open source community is suddenly not acknowledging a major, *fundamental issue* with their project which has a stable version, closing reported issues left and right!

**Reply by raveren** [reply]
I kept it from the comment, but 

> Personally I am holding back my immense disappointment after my whole team invested a huge amount of hours into a project relying heavily on spatie/data. At first we worked around the broken nested validation of laravel-data v1 and now v2 is just NOT allowing us to use validation AT ALL! I will never again pick up a library younger than 5 years for any business functionality, but for now I'm stuck and at least for the last two weeks have been struggling to validate the huge structural data we use to communicate with 3rd parties. I am backed into a corner here and would have dropped spatie/data a long time ago if I knew how much pain and work it would cause.

**Reply by freekmurze** [reply]
@raveren I'm sorry this package does not align with what you are expecting from it. Like always in open source, forking is an option. 

To keep maintainers and contributors motivated, please refrain from making negative non-technical comments (save those for online fora like Reddit).

>  I am backed into a corner

In open source, nobody ever forces you to pick a certain library or package. I'm happy that you are learning from it, and will be scrutinising open source more before landing on something.

**Reply by raveren** [reply]
I am sorry that you read negative, non-technical comment from what I wrote, I am sincerely trying to be as positive and optimistic as I can.

However, unless you're referring to the part which begins with "personally", and was ommitted as a side note to give you my background, I believe everything else is very technical!

**Reply by raveren** [reply]
I can work around `from()` not validating, even if it's surprising and you had people besides me be confused and write you about it (should I look it up?), but not having ANY way to validate a passed object is a blocker, guys, I am not negative, it just ain't possible!

**Reply by raveren** [reply]
I am really sorry that I seem offensive, I love and admire what you guys are doing!

**Reply by raveren** [reply]
Further to describe the situation from my side, how I stumbled upon the issue. 

My team wrote a huge `DtoMapper` functionality which is now used to convert objects to spatie dtos. One is supposed to create `class MyDtoToYourDtoMapper extends DtoMapper` and define how one object is converted into a destination DTO. There's a number of issues with it:
1. it has deep trouble when validating _and_ name mapping are both present in destination DTO - maybe the approach (convert to array then pass to destination `from()`) was not best, but I think this part didn't work out at all and the colleague had his contract end at this time :(
2. it is reproducing a LOT of what spatie/data already does - using reflection for public properties etc
3. mapper classes are difficult to discover/maintain and have no autocomplete of the destination columns
4. YOU GUYS HAVE A PERFECT SOLUTION ALREADY - the magic `from()` methods!

And the rest you know - turns out the magic `from()` methods are either not called or not validated :(

**Reply by raveren** [reply]
And again, final time, I promise, I really really struggle to find what it is that looks insulting (or condescending maybe??) in my tone :( I'm really sorry, I'm doing my best to just report the issue and all relevant aspects about it!

**Comment by adam-code-labx** [comment]
I came across a similar issue this afternoon using the from magic method.
`return DefaultResponse::from($this->model)->toArray();`

`ErrorException: Trying to access array offset on value of type null
at vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php:256
at vendor/spatie/laravel-data/src/Support/Creation/CreationContextFactory.php:47
at vendor/spatie/laravel-data/src/Concerns/BaseData.php:60
at vendor/spatie/laravel-data/src/Concerns/BaseData.php:46`

After a quick deep dive I could see the magic from method uses `CreationContextFactory::createFromConfig` which relies on the config. Now within a laravel application structure the config is resolved and there is no ErrorException.

However when using Laravel Data for packages you will need to include the LaravelDataServiceProvider which then resolves the data config.


```php
class TestCase extends Orchestra
{
    protected function setUp(): void
    {
        parent::setUp();

        $this->loadLaravelMigrations(['--database' => 'testbench']);
    }

    protected function getEnvironmentSetUp($app)
    {
        $app['config']->set('database.default', 'testbench');
        $app['config']->set('database.connections.testbench', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);
    }

    protected function getPackageProviders($app)
    {
        return [
            LaravelDataServiceProvider::class,
        ];
    }
}
```

@freekmurze Perhaps I should submit a PR for docs package testing here or perhaps a new Testing section under ADVANCED USAGE?
https://spatie.be/docs/laravel-data/v4/advanced-usage/performance#content-testing

**Reply by rubenvanassche** [reply]
If you want, you can add a new section to the advanced usage docs called "In Packages" and add a section about testing 👍

**Reply by adam-code-labx** [reply]
Thanks I will submit a PR this afternoon.

2nd step considering creating a PEST plugin as I am using data objects throughout my test suite and would love to make this more of a fluent API to streamline Data object testing for everyone 😜

**Reply by rubenvanassche** [reply]
Cool idea!
